components/table/table-wrapper.js

```diff
@@ -574,13 +579,34 @@ export class TableWrapper extends RtlMixin(PageableMixin(SelectionMixin(LitEleme
 		this._handleTableChange();
 	}
 
-	async _handleTableChange() {
+	async _handleTableChange(mutationRecords) {
+		const updates = { count: true, classNames: true, sticky: true, syncWidths: true };
+		if (mutationRecords) {
+			for (const key in updates) updates[key] = false;
+			for (const { type, removedNodes, addedNodes, target } of mutationRecords) {
+				if (type === 'attributes') {
+					updates.classNames = true;
+					continue;
+				}
+
+				updates.sticky ||= target.matches(SELECTORS.headers);
+				const affectedNodes = [...removedNodes, ...addedNodes];
+				for (const node of affectedNodes) {
+					if (!(node instanceof Element)) continue;
+					updates.classNames ||= node.matches('tr');
+					updates.syncWidths ||= node.matches('tr');
+					updates.sticky ||= node.matches(SELECTORS.headers);
+					updates.count ||= node.matches(SELECTORS.items);
+				}
+			}
+		}
+
 		await new Promise(resolve => requestAnimationFrame(resolve));
 
-		this._updateItemShowingCount();
-		this._applyClassNames();
-		this._syncColumnWidths();
-		this._updateStickyTops();
+		if (updates.count) this._updateItemShowingCount();
+		if (updates.classNames) this._applyClassNames();
+		if (updates.syncWidths) this._syncColumnWidths();
+		if (updates.sticky) this._updateStickyTops();
```

Comment:
This logic seems good to me! Just for my own clarity:

- Update the "item showing count" if any non-header rows are added/removed
- Update the class names if any row (header or otherwise) is added/removed or if the `selected` attribute is added/removed
- Sync the column widths if any row (header or otherwise) is added/removed. I its logic cares about cells here (`td` or `th`), although I'm not sure that in practice cells would get added/removed without also having rows added/removed. @grant-cleary FYI here.
- Update sticky if any header row is added/removed

Reply:
Cool stuff. There's a bug I noticed where on first render, the table doesn't end up syncing its widths properly. I meant to look into that, but I'm kind of curious if this just makes that problem go away. 🤔

Reply:
@dlockhart To add to your logic, if no records are given (i.e. not called from the `MutationObserver`) then all the updates run.
@grant-cleary I saw that too, but could not reproduce it with or without changes. But I don't think it's fixed, any update that we run now was ran before so whatever use-case is failing, it probably still is

Reply:
Gotcha, I'll dig more into that separately.